### DISCUSSION TITLE
Unexpected file size increase adding EXIF data

### DISCUSSION THREAD

**Discussion by ccalinescu** [SixLabors/ImageSharp#1696] [Q&A] [answered]
Hi there,

I am looking to use your library to add EXIF data to already compressed photos. I did couple of simple tests:

1. Open image and save without any transformation
`
            using (Image image = Image.Load(imagePath1)
            {
                             image.Save(imagePath2);
            }
`

2. Open image, add 2 keywords, save (no image transformation)
```
            using (Image image = Image.Load(imagePath1)
            {
                if (image.Metadata.ExifProfile == null)  image.Metadata.ExifProfile = new ExifProfile();

                image.Metadata.ExifProfile.SetValue(ExifTag.XPKeywords, Encoding.Unicode.GetBytes("hot, multiple"));

                             image.Save(imagePath2);
            }
```

Results for test 1 and 2 are the same:

- 220 KB JPG file size remains **same** at 220 KB
- 4 KB JPG file size **increase** to 8 KB (100% increase)
- 472 KB JPG file size **increase** to 532 KB (13% increase)

While I can understand that adding some info to a file will increase its size, the increase is not proportional with the data added and even if no data added the size increases.

Now I am in doubt if you use your library or not.

How do you explain that?
Is there any way to prevent or limit the file increase in size?

**Comment by tocsoft** [comment, marked as answer]
The ImageSharp pipeline is thus:

1. You call `Image.Load`
2. We read the entire image into memory, by decoding the compressed image format you provided in load 
3. We now have a fully decompressed set of pixel and metadata in memory (totally divorced from the source image data and compression)
4. _(Optionally) You make changes to the in memory image, be that the pixel or just metadata changes_
5. You call `image.Save()`
6. We load up a Format encoder (jpeg, png, bmp etc) based on the path you passed in above
7. We take the raw pixel and metadata and feed it into the the encoder along with our default (or what ever manually passed in set of) encoder configuration settings.
8. We convert the raw pixel and metadata using the config settings into an encoded image stream
9. We pipe the image data out (either to disk or memory)

As you will see from our processing pipeline the output data can easily end up with much at a different size even if you think you have done nothing to it. Just loading it creates a completely different data shape and data size than the encoded original, then we have the encode it again, but this time with a different encoder library with different settings than where was used when generating the source image.  

So in conclusion if you are hoping to use us for just manipulating EXIF data in jpegs without reencoding the pixel data then we probably are not the right choice. For resizing we might be, you just have to be conscious of the default quality settings (or equivalent, as 'quality' is really a JPEG only thing) of our various encoders and tune them until you are happy with the size/vs quality you end up with.

https://docs.sixlabors.com/articles/imagesharp/imageformats.html might be a good place to start looking for our various Formats

**Reply by ccalinescu** [reply]
Thank you for your quick answer.

I was thinking that something like this (a reencoding) may happen. From your explanation it seems there is no way to avoid this and I may need to experiment to see maybe what settings keep this resize in "reasonable" limits.

I hoped that you may have a setting/configuration that will allow to edit EXIF info without reencoding the file. Maybe a thing to consider for future releases.

**Comment by br3aker** [comment]
100% size increase (even 13% one) is not normal, can you provide test images before and after the conversion via ImageSharp?
